src/anomalib/metrics/per_image/pimo_numpy.py

```diff
@@ -0,0 +1,455 @@
+        msg = (
+            "The upper bound of the shared FPR integration range is not exactly achieved. "
+            f"Expected {fpr_upper_bound} but got {fpr_upper_bound_defacto}, which is not within {rtol=}."
+        )
+        warnings.warn(msg, RuntimeWarning, stacklevel=1)
+
+    # reminder: fpr lower/upper bound is threshold upper/lower bound (reversed)
+    thresh_lower_bound_idx = fpr_upper_bound_thresh_idx
+    thresh_upper_bound_idx = fpr_lower_bound_thresh_idx
+
+    # deal with edge cases
+    if thresh_lower_bound_idx >= thresh_upper_bound_idx:
+        msg = (
+            "The thresholds corresponding to the given `fpr_bounds` are not valid because "
+            "they matched the same threshold or the are in the wrong order. "
+            f"FPR upper/lower = threshold lower/upper = {thresh_lower_bound_idx} and {thresh_upper_bound_idx}."
+        )
+        raise RuntimeError(msg)
+
+    # limit the curves to the integration range [lbound, ubound]
+    shared_fpr_bounded: ndarray = shared_fpr[thresh_lower_bound_idx : (thresh_upper_bound_idx + 1)]
+    per_image_tprs_bounded: ndarray = per_image_tprs[:, thresh_lower_bound_idx : (thresh_upper_bound_idx + 1)]
+
+    # `shared_fpr` and `tprs` are in descending order; `flip()` reverts to ascending order
+    shared_fpr_bounded = np.flip(shared_fpr_bounded)
+    per_image_tprs_bounded = np.flip(per_image_tprs_bounded, axis=1)
+
+    # the log's base does not matter because it's a constant factor canceled by normalization factor
+    shared_fpr_bounded_log = np.log(shared_fpr_bounded)
+
+    # deal with edge cases
+    invalid_shared_fpr = ~np.isfinite(shared_fpr_bounded_log)
+
+    if invalid_shared_fpr.all():
+        msg = (
+            "Cannot compute AUPIMO because the shared fpr integration range is invalid). "
+            "Try increasing the number of thresholds."
+        )
+        raise RuntimeError(msg)
+
+    if invalid_shared_fpr.any():
+        msg = (
+            "Some values in the shared fpr integration range are nan. "
+            "The AUPIMO will be computed without these values."
+        )
+        warnings.warn(msg, RuntimeWarning, stacklevel=1)
+        logger.warning(msg)
+
+        # get rid of nan values by removing them from the integration range
+        shared_fpr_bounded_log = shared_fpr_bounded_log[~invalid_shared_fpr]
+        per_image_tprs_bounded = per_image_tprs_bounded[:, ~invalid_shared_fpr]
+
+    num_points_integral = int(shared_fpr_bounded_log.shape[0])
+
+    if num_points_integral <= 30:
+        msg = (
+            "Cannot compute AUPIMO because the shared fpr integration range doesnt have enough points. "
+            f"Found {num_points_integral} points in the integration range. "
+            "Try increasing `num_threshs`."
+        )
+        if not force:
+            raise RuntimeError(msg)
+        msg += " Computation was forced!"
+        warnings.warn(msg, RuntimeWarning, stacklevel=1)
```

Comment:
What's the advantage of using both logger.warning and warnings.warn. What happens if we want to pipe the warnings to a file and keep the console empty? A file handler can be passed to the logging module in such a scenario.

Reply:
tbh i didnt know what was the right policy here  😳
i put both cuz i was sure you'd find and tell me what to do :P hehe 

so should i use the logger?

on a side note: `warnings.warn` allows one to make this raise an exception with `warnings.filterwarnings("error")`; 
maybe that could be interesting? 

for context: this warning is due to having too few points to integrate the AUC curve, which may result in unprecise AUC values